okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java

```diff
@@ -320,9 +321,16 @@ private void connectTls(ConnectionSpecSelector connectionSpecSelector) throws IO
       SSLSession sslSocketSession = sslSocket.getSession();
       Handshake unverifiedHandshake = Handshake.get(sslSocketSession);
 
+      List<Certificate> peerCertificates = unverifiedHandshake.peerCertificates();
+
       // Verify that the socket's certificates are acceptable for the target host.
       if (!address.hostnameVerifier().verify(address.url().host(), sslSocketSession)) {
-        X509Certificate cert = (X509Certificate) unverifiedHandshake.peerCertificates().get(0);
+        if (peerCertificates.isEmpty()) {
+          throw new SSLPeerUnverifiedException(
+              "Hostname " + address.url().host() + " not verified, no certificates provided");
```

Comment:
Apparently there's a thing called “pre-shared keys” that make this case valid! Not sure how to check that though.

Reply:
Valid as in we should check if that is the case and allow those connections?  I'll look into it.